cnd/src/network/protocol.rs

```diff
@@ -0,0 +1,180 @@
+            let mut de = serde_json::Deserializer::from_slice(&message);
+            let info = SecretHashInfo::deserialize(&mut de)?;
+
+            Ok(info)
+        })
+    }
+}
+
+impl<C> OutboundUpgrade<C> for SecretHashProtocolConfig
+where
+    C: AsyncRead + AsyncWrite + Unpin + Send + Sync + 'static,
+{
+    type Output = ();
+    type Error = anyhow::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn upgrade_outbound(self, mut socket: C, _: Self::Info) -> Self::Future {
+        trace!("Upgrading outbound connection");
+        Box::pin(async move {
+            let message = serde_json::to_vec(&self.info)?;
+            upgrade::write_one(&mut socket, &message).await?;
+
+            Ok(())
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::swap_protocols::SwapId;
+    use futures::{channel::oneshot, prelude::*};
+    use libp2p::core::{
+        multiaddr::multiaddr,
+        transport::{memory::MemoryTransport, ListenerEvent, Transport},
+        upgrade::{apply_inbound, apply_outbound, Version},
+    };
+    use rand::{thread_rng, Rng};
+
+    static SWAP_ID: SwapId = SwapId::from_str("7f3a105d-ecf2-4cc6-b35c-b4351ac28a34").unwrap();
+
+    #[test]
+    fn swap_id_format_has_not_changed() {
+        let static_swap_id = SWAP_ID;
+        let default_swap_id = SwapId::default();
+
+        let static_swap_id = format!("{}", static_swap_id);
+        let default_swap_id = format!("{}", default_swap_id);
+
+        let want = static_swap_id.len();
+        let got = default_swap_id.len();
+
+        assert_eq!(want, got)
+    }
+
+    #[test]
+    fn correct_transfer() {
+        let bytes = b"hello world, you are beautiful!!";
+        let secret = Secret::from(*bytes);
+        let send_secret_hash = secret.hash();
+        let send_swap_id = SWAP_ID;
+
+        let (tx, rx) = oneshot::channel();
+
+        let mem_addr = multiaddr![Memory(thread_rng().gen::<u64>())];
+        let mut listener = MemoryTransport.listen_on(mem_addr).unwrap();
+
+        let listener_addr =
+            if let Some(Some(Ok(ListenerEvent::NewAddress(a)))) = listener.next().now_or_never() {
+                a
+            } else {
+                panic!("MemoryTransport not listening on an address!");
+            };
+
+        tokio::task::spawn(async move {
+            let listener_event = listener.next().await.unwrap();
+            let (listener_upgrade, _) = listener_event.unwrap().into_upgrade().unwrap();
+            let conn = listener_upgrade.await.unwrap();
+            let config = SecretHashProtocolConfig::sender(SecretHashInfo {
+                swap_id: send_swap_id,
+                secret_hash: send_secret_hash,
+            });
+            apply_outbound(conn, config, Version::V1).await.unwrap();
+        });
+
+        tokio::task::spawn_blocking(|| async move {
```

Comment:
I don't think `spawn_blocking` does what you think it is doing. It is meant for spawning blocking code, not blocking until it is done!

Reply:
https://docs.rs/tokio/0.2.13/tokio/task/fn.spawn_blocking.html

Reply:
oh, cheers.

Reply:
Take a look at the `tokio::test` attribute :)